src/commonMain/kotlin/fr/acinq/lightning/wire/LiquidityAds.kt

```diff
@@ -0,0 +1,153 @@
+package fr.acinq.lightning.wire
+
+import fr.acinq.bitcoin.*
+import fr.acinq.bitcoin.io.ByteArrayOutput
+import fr.acinq.bitcoin.io.Input
+import fr.acinq.bitcoin.io.Output
+import fr.acinq.lightning.MilliSatoshi
+import fr.acinq.lightning.blockchain.fee.FeeratePerKw
+import fr.acinq.lightning.channel.ChannelException
+import fr.acinq.lightning.channel.InvalidLiquidityAdsSig
+import fr.acinq.lightning.channel.LiquidityRatesRejected
+import fr.acinq.lightning.channel.MissingLiquidityAds
+import fr.acinq.lightning.transactions.Transactions
+import fr.acinq.lightning.utils.Either
+import fr.acinq.lightning.utils.msat
+import fr.acinq.lightning.utils.sat
+
+/**
+ * Liquidity ads create a decentralized market for channel liquidity.
+ * Nodes advertise fee rates for their available liquidity using the gossip protocol.
+ * Other nodes can then pay the advertised rate to get inbound liquidity allocated towards them.
+ */
+object LiquidityAds {
+
+    /**
+     * Liquidity is leased using the following rates:
+     *
+     *  - the buyer pays [[leaseFeeBase]] regardless of the amount contributed by the seller
+     *  - the buyer pays [[leaseFeeProportional]] (expressed in basis points) of the amount contributed by the seller
+     *  - the buyer refunds the on-chain fees for up to [[fundingWeight]] of the utxos contributed by the seller
+     *
+     * The seller promises that their relay fees towards the buyer will never exceed [[maxRelayFeeBase]] and [[maxRelayFeeProportional]].
+     * This cannot be enforced, but if the buyer notices the seller cheating, they should blacklist them and can prove
+     * that they misbehaved.
+     */
+    data class LeaseRates(val fundingWeight: Int, val leaseFeeProportional: Int, val maxRelayFeeProportional: Int, val leaseFeeBase: Satoshi, val maxRelayFeeBase: MilliSatoshi) {
+        val maxRelayFeeProportionalMillionths: Long = maxRelayFeeProportional.toLong() * 100
+
+        /**
+         * Fees paid by the liquidity buyer: the resulting amount must be added to the seller's output in the corresponding
+         * commitment transaction.
+         */
+        fun fees(feerate: FeeratePerKw, requestedAmount: Satoshi, contributedAmount: Satoshi): Satoshi {
```

Comment:
Integrators will find it useful to have a split between mining fees and "service" fees:

```suggestion
        data class LeaseFees(miningFee: Satoshi, serviceFee: Satoshi) {
          val fees: Satoshi = miningFee + serviceFee
        }

        fun fees(feerate: FeeratePerKw, requestedAmount: Satoshi, contributedAmount: Satoshi): LeaseFees {
```

Reply:
Good idea, done in https://github.com/ACINQ/lightning-kmp/pull/561/commits/c65c4cf919c3572095bc997868e8a01dbabb8ede